components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ServiceURLBuilder.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.core;
+
+import org.wso2.carbon.identity.core.internal.DefaultServiceURLBuilder;
+import org.wso2.carbon.identity.core.model.ServiceURL;
+
+/**
+ * URL Builder service interface.
+ */
+public interface ServiceURLBuilder {
+
+    static ServiceURLBuilder create() {
+
+        return new DefaultServiceURLBuilder();
+    }
+
+    /**
+     * This method is called to add the URL path to the builder.
+     *
+     * @param paths Context paths. Can provide multiple context paths with a comma separated string.
+     * @return {@link ServiceURLBuilder}.
+     */
+    ServiceURLBuilder addPath(String paths);
+
+    /**
+     * This method is called when the URL needs to be appended with the tenant domain as a query parameter when
+     * operating in legacy mode.
+     *
+     * @return {@link ServiceURLBuilder}.
+     */
+    ServiceURLBuilder addTenantAsQueryParam();
+
+    /**
+     * This method is called when the URL needs to be appended with the tenant domain as a path parameter when
+     * operating in legacy mode.
+     *
+     * @return {@link ServiceURLBuilder}.
+     */
+    ServiceURLBuilder addTenantAsPathParam();
+
+    /**
+     * This method is called to add the proxy context path to the URL.
+     *
+     * @return {@link ServiceURLBuilder}.
+     */
+    ServiceURLBuilder addProxyContext();
+
+    /**
+     * This method is called to add the web context root to the URL.
+     *
+     * @return {@link ServiceURLBuilder}.
+     */
+    ServiceURLBuilder addWebContext();
+
+    /**
+     * This method is called when the URL needs to be appended with other parameters. Such parameters should be
+     * entered as <k,v> pairs.
+     *
+     * @param key   Key.
+     * @param value Value.
+     * @return {@link ServiceURLBuilder}.
+     */
+    ServiceURLBuilder addParameter(String key, String value);
+
+    /**
+     * This method is called when the URL needs to be appended with a fragment.
+     *
+     * @param fragment Fragment.
+     * @return {@link ServiceURLBuilder}.
+     */
+    ServiceURLBuilder addFragment(String fragment);
```

Comment:
I'm not exactly sure if we should have key value supported as well. An overloaded method may be.
@mefarazath we are referring to key value pairs in oauth for fragments right

Reply:
From URL perspective there will be one fragment similar to query string. But we can have a method that accepts key, value pairs in a fragment, which might help building such fragments easily.

And above we can rename as setFragment which will denote that it will set the fragment component of the URL